indexer.go

```diff
@@ -0,0 +1,863 @@
+		toRemove += 1
+	}
+
+	cborData, err := EncodeListToCBOR(indexedFields)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Remove the padded 0s
+	unPadded := cborData[0 : len(cborData)-toRemove]
+
+
+	finalKey := Concat(indexPrefix, NULL_BYTE, unPadded)
+
+	return finalKey, nil
+}
+
+func (index Index) keyPrefix() ([]byte, error) {
+	nameBytes, err := IndexKeyFromFields(index.fields)
+	if err != nil {
+		return nil, err
+	}
+	return Concat(index.collection.keyPrefix(), INDEX_PREFIX, NULL_BYTE, nameBytes), nil
+}
+
+func (index Index) metadataKey() ([]byte, error) {
+	nameBytes, err := IndexKeyFromFields(index.fields)
+	if err != nil {
+		return nil, err
+	}
+	return Concat(index.collection.keyPrefix(), NULL_BYTE, INDEX_PREFIX, NULL_BYTE, nameBytes), nil
+}
+
+func (index Index) Exists() bool {
+	key, err := index.metadataKey()
+	if err != nil {
+		return false
+	}
+
+	_, err = index.collection.db.tree.Get(key)
+
+	return err == nil
+}
+
+func (index Index) Rebuild(ctx context.Context) error {
+	// Iterate over records in collection
+	// Insert index keys for each one
+	return nil
+}
+
+func (index Index) insert(ctx context.Context, record ipld.Node, recordId []byte) error {
+	indexRecordKey, err := index.recordKey(record, recordId)
+
+	if err != nil {
+		return err
+	}
+
+	ipldRecordId := basicnode.NewBytes(recordId)
+
+	return index.collection.db.tree.Put(ctx, indexRecordKey, ipldRecordId)
+}
+
+func (index Index) persistMetadata(ctx context.Context) error {
+	key, err := index.metadataKey()
+
+	if err != nil {
+		return err
+	}
+
+	metadata, err := qp.BuildMap(basicnode.Prototype.Any, -1, func(am datamodel.MapAssembler) {
+		qp.MapEntry(am, VERSION_KEY, qp.Int(INDEX_VERSION_1))
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return index.collection.db.tree.Put(ctx, key, metadata)
+}
+
+func (collection Collection) Iterate(ctx context.Context) (<-chan Record, error) {
+	prefix := collection.keyPrefix()
+	idStart := len(prefix) + len(DATA_PREFIX) + len(NULL_BYTE)
+	start := Concat(prefix, DATA_PREFIX, NULL_BYTE)
+	end := Concat(prefix, DATA_PREFIX, FULL_BYTE)
+
+	iterator, err := collection.db.tree.Search(ctx, start, end)
+
+	if err != nil {
+		return nil, err
+	}
+
+	linkSystem := collection.db.nodeStore.LinkSystem()
+
+	c := make(chan Record)
+
+	// TODO: Better error handling
+	go func(ch chan<- Record) {
+		defer close(c)
+		for !iterator.Done() {
+			// Ignore the key since we don't care about it
+			key, rawNode, err := iterator.NextPair()
+
+			if rawNode == nil && err == nil {
+				break
+			}
+
+			if err != nil {
+				panic(err)
+			}
+
+			cid, err := rawNode.AsLink()
+
+			if err != nil {
+				panic(err)
+			}
+
+			data, err := linkSystem.Load(
+				ipld.LinkContext{Ctx: ctx},
+				cid,
+				basicnode.Prototype.Map,
+			)
+
+			if err != nil {
+				panic(err)
+			}
+			id := key[idStart:]
+
+			record := Record{
+				Id:   id,
+				Data: data,
+			}
+
+			// TODO: What about the error?
```

Comment:
There are a few different ways of handling this: https://stackoverflow.com/questions/25142016/how-to-return-a-error-from-a-goroutine-through-channels

Reply:
Timeout, block the channel?